source/MRViewer/MRPlaneWidget.cpp

```diff
@@ -0,0 +1,171 @@
+    updateWidget_( updateCameraRotation );
+    if ( onPlaneUpdate_ )
+        onPlaneUpdate_();
+}
+
+void PlaneWidget::updateBox( const Box3f& box, bool updateCameraRotation )
+{
+    box_ = box;
+    updateWidget_( updateCameraRotation );
+}
+
+void PlaneWidget::definePlane()
+{
+    if ( planeObj_ )
+        return;
+
+    std::shared_ptr<Mesh> planeMesh = std::make_shared<Mesh>( makePlane() );
+    planeObj_ = std::make_shared<ObjectMesh>();
+    planeObj_->setName( "PlaneObject" );
+    planeObj_->setMesh( planeMesh );
+    planeObj_->setAncillary( true );
+    planeObj_->setFrontColor( Color( Vector4f::diagonal( 0.3f ) ), false );
+    planeObj_->setBackColor( Color( Vector4f::diagonal( 0.3f ) ) );
+    SceneRoot::get().addChild( planeObj_ );
+
+    updateWidget_();
+}
+
+void PlaneWidget::undefinePlane()
+{
+    if ( !planeObj_ )
+        return;
+    
+    planeObj_->detachFromParent();
+    planeObj_.reset();    
+}
+
+const Plane3f& PlaneWidget::getPlane() const
+{
+    return plane_;
+}
+
+const std::shared_ptr<ObjectMesh>& PlaneWidget::getPlaneObject() const
+{
+    return planeObj_;
+}
+
+void PlaneWidget::setOnPlaneUpdateCalback( OnPlaneUpdateCallback callback )
+{
+    onPlaneUpdate_ = callback;
+}
+
+void PlaneWidget::updateWidget_( bool updateCameraRotation )
+{
+    if ( !planeObj_ )
+        definePlane();
+
+    auto viewer = Viewer::instance();
+    plane_ = plane_.normalized();
+
+    auto trans1 = AffineXf3f::translation( plane_.project( box_.center() ) );
+    auto rot1 = AffineXf3f::linear( Matrix3f::rotation( Vector3f::plusZ(), plane_.n ) );
+    auto scale1 = AffineXf3f::linear( Matrix3f::scale( box_.diagonal() ) );
+    AffineXf3f transform = trans1 * rot1 * scale1;
+    if ( updateCameraRotation )
+        cameraUp3Old_ = viewer->viewport().getUpDirection();
+    Vector3f cameraUp3 = cameraUp3Old_;
+    auto rot2 = Matrix3f::rotation( transform.A * Vector3f::plusY(),
+                                    plane_.project( transform( Vector3f() ) + cameraUp3 ) - transform( Vector3f() ) );
+
+    auto lastPlaneTransform = trans1 * AffineXf3f::linear( rot2 ) * rot1;
+    transform = lastPlaneTransform * scale1;
+    planeObj_->setXf( transform );
+}
+
+bool PlaneWidget::onMouseDown_( Viewer::MouseButton button, int mod )
+{
+    if ( button != Viewer::MouseButton::Left || mod != 0 )
+        return false;
+
+    const auto& mousePos = Viewer::instance()->mouseController.getMousePos();
+    startMousePos_ = endMousePos_ = Vector2f( float( mousePos.x ), float( mousePos.y ) );
+    pressed_ = true;
+
+    line_ = std::make_shared<ObjectLines>();
+    line_->setName( "PlaneLine" );
+    line_->setAncillary( true );
+    const Polyline3 polyline( { { startMousePos_, endMousePos_ } } );
+    
+    const auto lineColor = SceneColors::get( SceneColors::Type::Labels );
+    line_->setFrontColor( lineColor, false );
+    line_->setBackColor( lineColor );
+
+    auto currentViewportId = Viewer::instance()->getHoveredViewportId();
```

Comment:
better use viewport().id here because line mouse can leave viewport and hovered id will be different from active viewport id